### ISSUE TITLE
Needed: easy way to setup service as plain HTTP (for development)

### ISSUE THREAD

**Issue by vivainio** [opensearch-project/OpenSearch#1598]
I would like to set up OpenSearch container image in development environments using plain HTTP.

This does not work out of the box, as the service defaults to HTTPS and crashes when plain old HTTP connection is attempted.

**Describe the solution you'd like**

In docker-compose.yml, I would like to be able to say "security.usehttp=true" or somesuch, and things should just work.

**Describe alternatives you've considered**

I created a custom docker image, but opensearch-dashboards doesn't work with that. Dockerfile:

```
FROM opensearchproject/opensearch:1.1.0
RUN /usr/share/opensearch/bin/opensearch-plugin remove opensearch-security
COPY --chown=opensearch:opensearch custom-opensearch.yml /usr/share/opensearch/config/
```

I also tried mounting custom opensearch.yml, but this is somewhat tricky in Windows docker and opensearch-dashboards didn't work with it either.

Environment variables that can be passed directly in docker-compose.yml are WAY easier than complete custom config file, especially since you won't easily break other functionality while changing a single flag.

**Comment by IMS94** [comment]
Similar concern is discussed in #1393

**Comment by setiah** [comment]
@vivainio you can disable security (hence HTTPS) via docker environment variables when running docker by setting `plugins.security.disabled=true`. Running the below command worked for me on Mac, (haven't tried windows). It automatically pulled the docker image and spinned up a cluster with http access. 

Reference: How to disable security - https://opensearch.org/docs/latest/security-plugin/configuration/disable/

```
docker run -p 9200:9200 -p 9600:9600 -e "discovery.type=single-node" -e "plugins.security.disabled=true" opensearchproject/opensearch:1.1.0
```

I agree with your larger point on making the overall experience simpler.

**Comment by namrataa-p** [comment]
@vivainio I had similar requirements and following docker-compose.yml worked for me on Windows.
```
version: '3'
services:
  opensearch-node1:
    image: opensearchproject/opensearch:1.2.0
    container_name: opensearch-node1
    environment:
      - cluster.name=opensearch-cluster
      - node.name=opensearch-node1
      - bootstrap.memory_lock=true
      - "OPENSEARCH_JAVA_OPTS=-Xms512m -Xmx512m"
      - "DISABLE_INSTALL_DEMO_CONFIG=true"
      - "DISABLE_SECURITY_PLUGIN=true"
      - "discovery.type=single-node"
    ulimits:
      memlock:
        soft: -1
        hard: -1
      nofile:
        soft: 65536
        hard: 65536
    volumes:
      - opensearch-data1:/usr/share/opensearch/data
    ports:
      - 9200:9200
      - 9600:9600
    networks:
      - opensearch-net
  opensearch-dashboards:
    image: opensearchproject/opensearch-dashboards:1.2.0
    container_name: opensearch-dashboards
    ports:
      - 5601:5601
    expose:
      - "5601"
    environment:
      - 'OPENSEARCH_HOSTS=["http://opensearch-node1:9200"]'
      - "DISABLE_SECURITY_DASHBOARDS_PLUGIN=true"
    networks:
      - opensearch-net
volumes:
  opensearch-data1:
networks:
  opensearch-net:
```

Found the `DISABLE_INSTALL_DEMO_CONFIG, DISABLE_SECURITY_PLUGIN, DISABLE_SECURITY_DASHBOARDS_PLUGIN` vars from [scenario-2-no-demo-certsconfigs--disable-security-on-both-opensearch-and-opensearch-dashboards](https://github.com/opensearch-project/opensearch-build/blob/main/docker/release/README.md#scenario-2-no-demo-certsconfigs--disable-security-on-both-opensearch-and-opensearch-dashboards)

**Comment by vivainio** [comment]
I can confirm that the environment variables (and docker-compose.yml) suggested by @namrataa-p work correctly. This ticket can be closed. I hope that this ticket will show up in google results for the next person having the problem :).

**Comment by dblock** [comment]
Before we close this, shouldn't we publish this somewhere in a repo and/or add to [the documentation](https://github.com/opensearch-project/documentation-website)?

**Comment by namrataa-p** [comment]
@dblock the `DISABLE_SECURITY_PLUGIN, DISABLE_SECURITY_DASHBOARDS_PLUGIN` environment variables should be part of the documentation at the least. 
I would have liked it if this info was present under [docker](https://github.com/opensearch-project/documentation-website/blob/main/_opensearch/install/docker.md)/[docker-security](https://github.com/opensearch-project/documentation-website/blob/main/_opensearch/install/docker-security.md) sections in the documentation.

If you agree with this then maybe I could try to add this info to one of the appropriate sections.

**Comment by vivainio** [comment]
It would be best if these options were in commented out environment blocks in sample docker compose file: https://opensearch.org/docs/latest/opensearch/install/docker/#sample-docker-compose-file

**Comment by Babas-1** [comment]
Yes please! I spent many hours digging for this solution. Please add it to the docker compose file so others do not have to.

**Comment by dblock** [comment]
> @dblock the `DISABLE_SECURITY_PLUGIN, DISABLE_SECURITY_DASHBOARDS_PLUGIN` environment variables should be part of the documentation at the least. I would have liked it if this info was present under [docker](https://github.com/opensearch-project/documentation-website/blob/main/_opensearch/install/docker.md)/[docker-security](https://github.com/opensearch-project/documentation-website/blob/main/_opensearch/install/docker-security.md) sections in the documentation.
> 
> If you agree with this then maybe I could try to add this info to one of the appropriate sections.

yes please, anyone can do it - @vivainio and @Babas-1 too

**Comment by Babas-1** [comment]
It might be good to also have logstash in the docker compose file or mention in the documentation how to connect it to the docker image that the compose file initiates.

**Comment by elfisher** [comment]
> It might be good to also have logstash in the docker compose file or mention in the documentation how to connect it to the docker image that the compose file initiates.

Agreed. It would be nice to have an optional compose that includes Logstash, FluentD/Bit, or Data Prepper, etc... Let's track that as a separate request and discuss it further.

**Comment by minalsha** [comment]
For this issue specifically, we have the PR:https://github.com/opensearch-project/documentation-website/pull/310 addressing the documentation update and hence closing the issue.

**Comment by hgati** [comment]
> @vivainio you can disable security (hence HTTPS) via docker environment variables when running docker by setting `plugins.security.disabled=true`. Running the below command worked for me on Mac, (haven't tried windows). It automatically pulled the docker image and spinned up a cluster with http access.
> 
> Reference: How to disable security - https://opensearch.org/docs/latest/security-plugin/configuration/disable/
> 
> ```
> docker run -p 9200:9200 -p 9600:9600 -e "discovery.type=single-node" -e "plugins.security.disabled=true" opensearchproject/opensearch:1.1.0
> ```
> 
> I agree with your larger point on making the overall experience simpler.
> 
> cc: @elfisher

`plugins.security.disabled=true`.
My issue is solved. thanks.

**Comment by dss010101** [comment]
> ut environment blocks in sample docker compose file:

i feel like this kind of reasoning is why i've spent a week trying to get a simple example of open-search working.  documentation just blows...none of the sample compose work out of the box...and it's extremely difficult to find help.

**Comment by dblock** [comment]
I feel your pain @dss010101. "i've spent a week trying to get a simple example of open-search working" is clearly not acceptable, so let me try to help.

First, re:"documentation just blows" and "none of the sample compose work out of the box" - this is clearly a problem and we should be able to fix it, shouldn't we? I see folks in the issue above actually doing something about it, opening issues, making fixes. Thanks! 

We weren't able to take with us the Elastic docs after the fork, so all that had to be (re)written from scratch. I guess I'm saying we need help! For your use-case all it takes is one motivated person to go through docs, get things working from scratch and correct/update/improve the documentation till a 2yo can set things up.

Stepping back. What is your next problem and what are you blocked on?

**Comment by dss010101** [comment]
> 
Well i started with this but could not get any to work: https://github.com/opensearch-project/data-prepper/tree/main/examples/log-ingestion.  There were various errors, but it appears that OS did not like fluentbit trying to communicate via ssl/tls, which i found odd (es on single node configures security for you).
I tried a few other docker-compose examples out there - none worked out of the box.

I was getting errors like these: 
`[o.o.s.s.h.n.SecuritySSLNettyHttpServerTransport] [55811de536b0] Exception during establishing a SSL connection: io.netty.handler.ssl.NotSslRecordException: not an SSL/TLS record`

Eventually, through mostly FluentBit docs, i learned i could turn that security off..so i did.  and that got me further.
my fluentbit.conf + parsers.conf defined a multiline parser.   OS then complained it did not like the 'multiline-regex-test' i had setup, even though i was defined in a parsers.conf file and with fluentbit Output set to stdout,  worked fine.  So i replaced that with just the fluenbit built in python parser, which it did like.  
But then it complained about some 'type' argument. so i found through FluentBit docs again i could turn that (which i dont really know what it is tbh) off.   Eventually ended up with this configuration that worked with out SSL that looked like this

[SERVICE]
  Flush                 1
  Log_Level             debug
  parsers_file          parsers.conf
  trace_error           On

[INPUT]
  name                  tail
  tag                   rasp_svc
  refresh_interval      5
  path                  /var/log/test.log
  path_key              file_name
  read_from_head        true
  multiline.parser      python

[OUTPUT]
  Name opensearch
  Match *
  host opensearch
  Replace_Dots On
  tls Off
  tls.verify Off
  Suppress_Type_Name On
```

so not logs are being fed to ES.  except now also, the Dashboard stopped working..i cant log into it because apparently security is not setup.  so, im close to throwing my hands up in the air...thinking maybe Graffana + Loki might be a better choice for me.
Primarily because not having security isnt an option and if it took this long and this much pain to get a simple example going  - i cant imagine the nightmare that awaits me when i try to properly setup ssl/certififates.

**Comment by dblock** [comment]
> I was getting errors like these:
> `[o.o.s.s.h.n.SecuritySSLNettyHttpServerTransport] [55811de536b0] Exception during establishing a SSL connection: io.netty.handler.ssl.NotSslRecordException: not an SSL/TLS record`

This meant that something was trying to do a non-SSL request to an SSL endpoint, as expected.

It sounds like you just aren't able to get fluentbit to work with OpenSearch with SSL enabled?

**Comment by dss010101** [comment]
> > I was getting errors like these:
> > `[o.o.s.s.h.n.SecuritySSLNettyHttpServerTransport] [55811de536b0] Exception during establishing a SSL connection: io.netty.handler.ssl.NotSslRecordException: not an SSL/TLS record`
> 
> This meant that something was trying to do a non-SSL request to an SSL endpoint, as expected.
> 
> It sounds like you just aren't able to get fluentbit to work with OpenSearch with SSL enabled?

Yes, that's right i believe.  the problem is if i set 'tls on' in fluenbit.conf: i then get this error from OS:

`opensearch             | [2023-05-12T22:44:21,317][ERROR][o.o.s.s.h.n.SecuritySSLNettyHttpServerTransport] [75b97ad77d15] Exception during establishing a SSL connection: javax.net.ssl.SSLHandshakeException: Insufficient buffer remaining for AEAD cipher fragment (2). Needs to be more than tag size (16)`

**Comment by dblock** [comment]
@dss010101 one of the root causes of this may be https://bugs.openjdk.org/browse/JDK-8221218. Which JDK are you using?

**Comment by dblock** [comment]
Btw, `Suppress_Type_Name` is documented in https://docs.fluentbit.io/manual/pipeline/outputs/elasticsearch as "When enabled, mapping types is removed and Type option is ignored. Types are deprecated in APIs in [v7.0](https://www.elastic.co/guide/en/elasticsearch/reference/current/removal-of-types.html). This options is for v7.0 or later.", it's the same in OpenSearch 2.x, mapping types were deprecated in ES 7.x and removed in OpenSearch 2.x (https://opensearch.org/docs/2.0/breaking-changes/#remove-mapping-types-parameter).

**Comment by dss010101** [comment]
> 

hi..i've moved on from OS.  But i was basically pulling latest image for each of the stack component.
And yes, i did come across that reference to suppress_type_name in the docs. thanks.

**Comment by sbmpost** [comment]
@dss010101 

> i feel like this kind of reasoning is why i've spent a week trying to get a simple example of open-search working. documentation just blows...none of the sample compose work out of the box...and it's extremely difficult to find help.

Similar to you I just spent 3 days getting things to work locally. Following the instructions of the [documentation](https://opensearch.org/docs/1.1/security-plugin/configuration/disable/) simply didn't work. This is certainly not a warm welcome when you are new to opensearch (which I am). For what is worth, here are the commands that I now run to quickly get something running and allows access from the brower using http://localhost:9200 for opensearch and http://localhost/5601 for the dashboards.

    terminal1: docker network create opensearch

    terminal1: docker run --rm --restart=no --net opensearch -p 9200:9200 -e cluster.name="opensearch-cluster" -e node.name="os01" -e discovery.type="single-node" -e bootstrap.memory_lock="true" -e plugins.security.disabled="true" --name=os-node1 docker.io/opensearchproject/opensearch:2.7.0

    terminal2: docker run --rm --restart=no --net opensearch -p 5601:5601 -e server.name="opensearch_dashboards" -e server.host="localhost" -e DISABLE_SECURITY_DASHBOARDS_PLUGIN="true" -e OPENSEARCH_HOSTS="http://$(docker inspect os-node1 | jq -r .[].NetworkSettings.Networks.opensearch.IPAddress):9200" --name=os-dashboards opensearch-dashboards:2.7.0

Note the terminal2 command uses 'jq' to grep the ipaddress from the docker instance started in terminal1. My next challenge will be to get rid of these errors in the opensearch logs (running on OSX):

    [2023-05-17T11:37:30,687][WARN ][o.o.b.JNANatives         ] [os01] Unable to lock JVM Memory: error=12, reason=Cannot allocate memory
    [2023-05-17T11:37:30,698][WARN ][o.o.b.JNANatives         ] [os01] This can result in part of the JVM being swapped out.
    [2023-05-17T11:37:30,698][WARN ][o.o.b.JNANatives         ] [os01] Increase RLIMIT_MEMLOCK, soft limit: 8388608, hard limit: 8388608
    [2023-05-17T11:37:30,699][WARN ][o.o.b.JNANatives         ] [os01] These can be adjusted by modifying /etc/security/limits.conf, for example

Will report back in probably another week to get these fixed...

**Comment by dtaivpp** [comment]
@dblock yeah we are actually working on re-doing the quickstart experience (as well as some of the security on boarding). Where do you think we should track the progress? We are going to be touching a lot on documentation and security. Maybe we should start with the meta issue in the security repo?

**Comment by dblock** [comment]
> Maybe we should start with the meta issue in the security repo?

For a developer quickstart specifically with security that makes sense. Please create and link it here!

**Comment by dss010101** [comment]
FYI - google had taken me to another project where a similar issue was seen.  the owner of that project responded.  Just passing this on incase it may be helpful to others:

https://github.com/tkohegyi/mitmJavaProxy/issues/12#issuecomment-1554253334

**Comment by kishykumar** [comment]
Can this be added to the official docker opensearch page: https://hub.docker.com/r/opensearchproject/opensearch please?
I spent many hours digging this.

**Comment by dblock** [comment]
@kishykumar Yes, care to contribute please?

**Comment by kishykumar** [comment]
@dblock - I am not sure how to contribute. Can you please let me know how can I modify or submit a modification to this page: https://hub.docker.com/r/opensearchproject/opensearch?

**Comment by kishykumar** [comment]
For the benefit of others, the following command worked for me:

`$ docker run -it -p 9200:9200 -p 9600:9600 -e OPENSEARCH_INITIAL_ADMIN_PASSWORD="Pass#word#10" -e "discovery.type=single-node"  -e "plugins.security.disabled=true" --name opensearch-node opensearchproject/opensearch:latest`

The trick is adding:  `-e "plugins.security.disabled=true"`

**Comment by dblock** [comment]
@peterzhuamazon do you know? ^

**Comment by peterzhuamazon** [comment]
> @peterzhuamazon do you know? ^

Yes, but why do we disable security after all?
Sorry I am a bit confused here.

Thanks.

**Comment by peterzhuamazon** [comment]
Also in order to disable security on docker you can follow the parameters here:
https://github.com/opensearch-project/opensearch-build/blob/main/docker/release/README.md
